File: Leetcode Daily Challenge/016-October-2024/028-longest-square-streak-in-an-array.cpp
// https://leetcode.com/problems/longest-square-streak-in-an-array/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // TC : O(N)
    // SC : O(N)
    int longestSquareStreak(vector<int>& nums) {
        unordered_set<int> st(begin(nums), end(nums));
        int result = 0;
        for (int &num : nums) {
            int subLen = 0;
            long currNum = num;
            while (currNum <= 1e5 && st.count(currNum)) {   // O(5) at worst : 2, 4, 16, 256, 65536, 4294967296
                subLen ++;
                currNum *= currNum;
            }
            result = max(result, subLen);
        }
        return result < 2 ? -1 : result;
    }
};

File: Leetcode Daily Challenge/016-October-2024/028-longest-square-streak-in-an-array.java
// https://leetcode.com/problems/longest-square-streak-in-an-array/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(N)
    public int longestSquareStreak(int[] nums) {
        HashSet<Integer> st = new HashSet<>();
        for (int num : nums) {
            st.add(num);
        }
        int maxStreak = 0;
        for (int num : nums) {
            long currNum = num;
            int subqLen = 0;
            while (currNum <= 1e5 && st.contains((int)currNum)) {   // At Worst : O(5)
                subqLen += 1;
                currNum *= currNum;
            }
            maxStreak = Math.max(maxStreak, subqLen);
        }
        return maxStreak < 2 ? - 1 : maxStreak;
    }
}